external/storm-pmml/src/main/java/org/apache/storm/pmml/PMMLPredictorBolt.java

```diff
@@ -0,0 +1,118 @@
+
+import java.util.List;
+import java.util.Map;
+
+public class PMMLPredictorBolt extends BaseRichBolt {
+    protected static final Logger LOG = LoggerFactory.getLogger(PMMLPredictorBolt.class);
+
+    private ModelOutputFields outFields;
+    private ModelRunnerFactory runnerFactory;
+    private ModelRunner runner;
+    private OutputCollector collector;
+
+    /*
+     * Passing a factory rather than the actual object to avoid enforcing the strong
+     * requirement of having to have ModelRunner to be Serializable
+     */
+
+    /**
+     * Creates an instance of {@link PMMLPredictorBolt} that executes, for every tuple, the runner constructed with
+     * the {@link ModelRunnerFactory} specified in the parameter
+     * The {@link PMMLPredictorBolt} instantiated with this constructor has no declared output fields
+     */
+    public PMMLPredictorBolt(ModelRunnerFactory modelRunnerFactory) {
+        this(modelRunnerFactory, null);
+    }
+
+    /**
+     * Creates an instance of {@link PMMLPredictorBolt} that executes, for every tuple, the runner constructed with
+     * the {@link ModelRunnerFactory} specified in the parameter
+     * The {@link PMMLPredictorBolt} instantiated with this constructor declares the output fields as specified
+     * by the {@link ModelOutputFields} parameter
+     */
+    public PMMLPredictorBolt(ModelRunnerFactory modelRunnerFactory, ModelOutputFields modelOutputFields) {
+        this.outFields = modelOutputFields;
+        this.runnerFactory = modelRunnerFactory;
+        LOG.info("Instantiated {}", this);
+    }
+
+    @Override
+    public void prepare(Map stormConf, TopologyContext context, OutputCollector collector) {
+        this.runner = runnerFactory.newModelRunner();
+        this.collector = collector;
+    }
+
+    @Override
+    public void execute(Tuple input) {
+        try {
+            final Map<Stream, List<Object>> scoresPerStream = runner.scoredTuplePerStream(input);
+            LOG.debug("Input tuple [{}] generated predicted scores [{}]", input, scoresPerStream);
+            if (scoresPerStream != null) {
+                for (Stream stream : scoresPerStream.keySet()) {
+                    collector.emit(stream.getId(), input, scoresPerStream.get(stream));
+                }
+            } else {
+                LOG.debug("Input tuple [{}] generated NULL scores", input);
+            }
+        } catch(Exception e) {
+            collector.reportError(e);
+            collector.fail(input);
+        }
+    }
+
+    @Override
+    public void declareOutputFields(OutputFieldsDeclarer declarer) {
+        if (outFields != null) {
+            LOG.info("Declaring output fields [{}]", outFields);
+            for (Map.Entry<? extends Stream, ? extends Fields> of : outFields.toMap().entrySet()) {
+                final Stream stream = of.getKey();
+                final Fields fields = of.getValue();
+                declarer.declareStream(stream.getId(), stream.isDirect(), fields);
```

Comment:
doesnt make sense to declare stream as direct unless you intend to emit directly to specific tasks via emitDirect.